File: strings/3455-shortest-matching-substring.go
package strings

import (
	"math"
	"sort"
	"strings"
)

// https://leetcode.com/problems/shortest-matching-substring/

// Approach: Binary Search with Pattern Matching
// Find all occurrences of prefix, middle, suffix, then use binary search
// to find the shortest valid combination
// Time: O(n log n) where n = len(s)
// Space: O(n) for storing occurrences
//
//nolint:gocyclo,revive,cyclop
func minValidSubstring(s string, p string) int {
	// Split pattern by '*'
	parts := strings.Split(p, "*")
	if len(parts) != 3 {
		return -1
	}

	prefix, middle, suffix := parts[0], parts[1], parts[2]

	// Edge case: empty pattern parts
	if prefix == "" && middle == "" && suffix == "" {
		return 0
	}

	lenPre := len(prefix)
	lenMid := len(middle)
	lenSuf := len(suffix)

	// Find all occurrences of a substring
	findAllOccurrences := func(part string) []int {
		if part == "" {
			return []int{}
		}

		positions := []int{}
		pos := 0

		for {
			idx := strings.Index(s[pos:], part)
			if idx == -1 {
				break
			}

			positions = append(positions, pos+idx)
			pos = pos + idx + 1
		}

		return positions
	}

	listP := findAllOccurrences(prefix)
	listM := findAllOccurrences(middle)
	listS := findAllOccurrences(suffix)

	minLen := math.MaxInt32

	// Handle all 8 cases based on which parts are empty
	if prefix != "" && middle != "" && suffix != "" {
		// All three parts present
		for _, pStart := range listP {
			pEnd := pStart + lenPre
			// Find first middle at or after prefix ends
			mIdx := sort.SearchInts(listM, pEnd)
			if mIdx >= len(listM) {
				continue
			}

			mStart := listM[mIdx]
			mEnd := mStart + lenMid
			// Find first suffix at or after middle ends
			sIdx := sort.SearchInts(listS, mEnd)
			if sIdx >= len(listS) {
				continue
			}

			sStart := listS[sIdx]
			sEnd := sStart + lenSuf

			length := sEnd - pStart
			if length < minLen {
				minLen = length
			}
		}
	} else if prefix == "" && middle != "" && suffix != "" {
		// Only middle and suffix
		for _, mStart := range listM {
			mEnd := mStart + lenMid

			sIdx := sort.SearchInts(listS, mEnd)
			if sIdx >= len(listS) {
				continue
			}

			sStart := listS[sIdx]
			sEnd := sStart + lenSuf

			length := sEnd - mStart
			if length < minLen {
				minLen = length
			}
		}
	} else if prefix != "" && middle == "" && suffix != "" {
		// Only prefix and suffix
		for _, pStart := range listP {
			pEnd := pStart + lenPre

			sIdx := sort.SearchInts(listS, pEnd)
			if sIdx >= len(listS) {
				continue
			}

			sStart := listS[sIdx]
			sEnd := sStart + lenSuf

			length := sEnd - pStart
			if length < minLen {
				minLen = length
			}
		}
	} else if prefix != "" && middle != "" && suffix == "" {
		// Only prefix and middle
		for _, pStart := range listP {
			pEnd := pStart + lenPre

			mIdx := sort.SearchInts(listM, pEnd)
			if mIdx >= len(listM) {
				continue
			}

			mStart := listM[mIdx]
			mEnd := mStart + lenMid

			length := mEnd - pStart
			if length < minLen {
				minLen = length
			}
		}
	} else if prefix != "" && middle == "" && suffix == "" {
		// Only prefix (pattern: "prefix**")
		if len(listP) > 0 {
			minLen = lenPre
		}
	} else if prefix == "" && middle != "" && suffix == "" {
		// Only middle (pattern: "*middle*")
		if len(listM) > 0 {
			minLen = lenMid
		}
	} else if prefix == "" && middle == "" && suffix != "" {
		// Only suffix (pattern: "**suffix")
		if len(listS) > 0 {
			minLen = lenSuf
		}
	}

	if minLen == math.MaxInt32 {
		return -1
	}

	return minLen
}

// // Approach #2: Pattern matching without binary search (TLE)
// // Time: O(n^2) where n = len(s)
// // Space: O(1)
// func minValidSubstring(s string, p string) int {
// 	indexFrom := func(s, substr string, start int) int {
// 		if start >= len(s) {
// 			return -1
// 		}
//
// 		for i := start; i <= len(s)-len(substr); i++ {
// 			if s[i:i+len(substr)] == substr {
// 				return i
// 			}
// 		}
//
// 		return -1
// 	}
//
// 	// Split pattern by '*'
// 	parts := make([]string, 0, 3)
// 	current := ""
//
// 	for _, ch := range p {
// 		if ch == '*' {
// 			parts = append(parts, current)
// 			current = ""
// 		} else {
// 			current += string(ch)
// 		}
// 	}
//
// 	parts = append(parts, current)
//
// 	if len(parts) != 3 {
// 		return -1
// 	}
//
// 	prefix, middle, suffix := parts[0], parts[1], parts[2]
//
// 	// Edge case: empty pattern parts (like "**")
// 	if prefix == "" && middle == "" && suffix == "" {
// 		return 0
// 	}
//
// 	n := len(s)
// 	minLen := math.MaxInt32
//
// 	// Try all possible starting positions
// 	for start := range n {
// 		// Match prefix
// 		pos := start
// 		if prefix != "" {
// 			if start+len(prefix) > n {
// 				continue
// 			}
//
// 			if s[start:start+len(prefix)] != prefix {
// 				continue
// 			}
//
// 			pos = start + len(prefix)
// 		}
//
// 		// Try to find middle starting from pos
// 		if middle != "" {
// 			midPos := indexFrom(s, middle, pos)
// 			if midPos == -1 {
// 				continue
// 			}
//
// 			pos = midPos + len(middle)
// 		}
//
// 		// Try to find suffix starting from pos
// 		var end int
//
// 		if suffix != "" {
// 			sufPos := indexFrom(s, suffix, pos)
// 			if sufPos == -1 {
// 				continue
// 			}
//
// 			end = sufPos + len(suffix)
// 		} else {
// 			end = pos
// 		}
//
// 		// Found a match
// 		length := end - start
// 		if length < minLen {
// 			minLen = length
// 		}
// 	}
//
// 	if minLen == math.MaxInt32 {
// 		return -1
// 	}
//
// 	return minLen
// }

File: strings/3455-shortest-matching-substring.py
from bisect import bisect_left, bisect_right
import unittest
from typing import List

# https://leetcode.com/problems/shortest-matching-substring/

# python3 -m unittest strings/3455-shortest-matching-substring.py


class Solution(unittest.TestCase):
    # # Approach #1: Pattern matching without binary search (TLE)
    # # Time: O(n^2) where n = len(s)
    # # Space: O(1)
    # def minValidSubstring(self, s: str, p: str) -> int:
    #     parts = p.split("*")
    #     if len(parts) != 3:
    #         return -1

    #     prefix, middle, suffix = parts

    #     if not prefix and not middle and not suffix:
    #         return 0

    #     n = len(s)
    #     min_len = float("inf")

    #     for start in range(n):
    #         if prefix:
    #             if start + len(prefix) > n:
    #                 continue
    #             if s[start : start + len(prefix)] != prefix:
    #                 continue
    #             pos = start + len(prefix)
    #         else:
    #             pos = start

    #         if middle:
    #             mid_pos = s.find(middle, pos)
    #             if mid_pos == -1:
    #                 continue
    #             pos = mid_pos + len(middle)

    #         if suffix:
    #             suf_pos = s.find(suffix, pos)
    #             if suf_pos == -1:
    #                 continue
    #             end = suf_pos + len(suffix)
    #         else:
    #             end = pos

    #         min_len = min(min_len, end - start)

    #     return int(min_len) if min_len != float("inf") else -1

    # Approach #2: Binary Search with Pattern Matching
    # Time: O(n log n) where n = len(s)
    # Space: O(n) for storing occurrences
    def minValidSubstring(self, s: str, p: str) -> int:
        # Split pattern into A*B*C
        parts = p.split('*')
        A, B, C = parts[0], parts[1], parts[2]
        a, b, c = len(A), len(B), len(C)

        # KMP to find all occurrences of a pattern in text
        def kmp_occurrences(text: str, pat: str) -> List[int]:
            m = len(pat)
            if m == 0:
                return []  # we won't call this for empty patterns
            # Build LPS
            lps = [0] * m
            j = 0
            for i in range(1, m):
                while j > 0 and pat[i] != pat[j]:
                    j = lps[j - 1]
                if pat[i] == pat[j]:
                    j += 1
                    lps[i] = j
            # Search
            res: List[int] = []
            j = 0
            for i in range(len(text)):
                while j > 0 and text[i] != pat[j]:
                    j = lps[j - 1]
                if text[i] == pat[j]:
                    j += 1
                    if j == m:
                        res.append(i - m + 1)
                        j = lps[j - 1]
            return res

        # Occurrences lists for non-empty tokens
        posA = kmp_occurrences(s, A) if a > 0 else []
        posB = kmp_occurrences(s, B) if b > 0 else []
        posC = kmp_occurrences(s, C) if c > 0 else []

        # Early impossibility checks
        if a > 0 and not posA:
            return -1
        if b > 0 and not posB:
            return -1
        if c > 0 and not posC:
            return -1

        INF = 10**18
        ans = INF

        # Case breakdown

        # Case A empty, C empty
        if a == 0 and c == 0:
            if b == 0:
                return 0  # "**"
            else:
                return b if posB else -1  # "*B*"

        # Case A non-empty, C empty
        if a > 0 and c == 0:
            if b == 0:
                # Pattern A* -> shortest is just A if it exists
                return a
            else:
                # For each A start, take earliest B at/after i+a
                for i in posA:
                    idxB = bisect_left(posB, i + a)
                    if idxB == len(posB):
                        continue
                    bStart = posB[idxB]
                    cand = bStart + b - i
                    if cand < ans:
                        ans = cand
                return ans if ans < INF else -1

        # Case A empty, C non-empty
        if a == 0 and c > 0:
            if b == 0:
                # Pattern *C -> shortest is just C if it exists
                return c
            else:
                # For each C start, pick latest B start with bStart + b <= cStart
                for cStart in posC:
                    xMax = cStart - b
                    if xMax < 0:
                        continue
                    idxB = bisect_right(posB, xMax) - 1
                    if idxB < 0:
                        continue
                    bStart = posB[idxB]
                    cand = cStart + c - bStart
                    if cand < ans:
                        ans = cand
                return ans if ans < INF else -1

        # Case A non-empty, C non-empty
        # General form: A * B * C
        if b == 0:
            # Reduce to A * C: for each A start, take earliest C start >= i + a
            for i in posA:
                idxC = bisect_left(posC, i + a)
                if idxC == len(posC):
                    continue
                cStart = posC[idxC]
                cand = cStart + c - i
                if cand < ans:
                    ans = cand
            return ans if ans < INF else -1
        else:
            # For each A start, take earliest B at/after i+a, then earliest C at/after bStart+b
            for i in posA:
                idxB = bisect_left(posB, i + a)
                if idxB == len(posB):
                    continue
                bStart = posB[idxB]
                y = bStart + b
                idxC = bisect_left(posC, y)
                if idxC == len(posC):
                    continue
                cStart = posC[idxC]
                cand = cStart + c - i
                if cand < ans:
                    ans = cand
            return ans if ans < INF else -1

    def test(self):
        for s, pattern, expected in [
            # Valid test cases (pattern has exactly two '*')
            ("abaacbaecebce", "ba*c*ce", 8),
            ("baccbaadbc", "cc*baa*adb", -1),
            ("a", "**", 0),
            ("madlogic", "*adlogi*", 6),
            ("hy", "h**", 1),
        ]:
            output = self.minValidSubstring(s, pattern)
            self.assertEqual(expected, output, f"expected: {expected}, output: {output}")
